src/cargo/util/graph.rs

```diff
@@ -1,52 +1,227 @@
+use indexmap::IndexMap;
 use std::borrow::Borrow;
-use std::collections::BTreeSet;
+use std::cell::{Ref, RefCell, RefMut};
+use std::collections::HashSet;
 use std::fmt;
+use std::hash::Hash;
+use std::num::NonZeroUsize;
+use std::rc::Rc;
 
-use im_rc;
+#[derive(Clone, Debug)]
+struct EdgeLink<E: Clone> {
+    value: E,
+    next: Option<NonZeroUsize>,
+}
 
+/// This is a directed Graph structure. Each edge can have an `E` associated with it,
+/// but may have more then one or none. Furthermore, it is designed to be "append only" so that
+/// it can be queried as it would have bean when it was smaller. This allows a `reset_to` method
+/// that efficiently undoes the most reason modifications.
```

Comment:
Could the comments here and the fields below be expanded to indicate what all the integer pointers are intended for? From reading this I don't have a great grasp of how this is implemented, and it looks pretty complicated internally. To clarify I don't doubt it, was just hoping to not have to read so closely below after reading this :)

Reply:
I will add a new type for each use of `usize` and some comments.

A fair bit of the complexity comes from the `or none` support. This is needed as a lockfile knows all the packages eth package depends on but does not know the real `Dependency` objects. This data structure supports interweaving of adding links with edge data and links that do not have data. We use both but we don't in practice interweave them. I wonder if there is a way to simplify the structure if we ban interweaving...